### DISCUSSION TITLE
[QST] Some confusion for GpuSemaphore.releaseIfNecessary()

### DISCUSSION THREAD

**Discussion by JustPlay** [NVIDIA/spark-rapids#5385] [General] [answered]
https://github.com/NVIDIA/spark-rapids/blame/1fa70c3aa36833d751b07390959896f0a2924e74/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala#L128-L142

For `GpuSemaphore.acquireIfNecessary()` and `GpuSemaphore.completeTask()` I know exactly what they are doing, but for `GpuSemaphore.releaseIfNecessary()` i'am a little confused. 

I think, `GpuSemaphore.releaseIfNecessary()` should release it directly, not just dec the reference count (and rm it only when ref == 0). When a node call `GpuSemaphore.releaseIfNecessary()` it means the node itself will leave the gpu and all the data related to this node and the upstream nodes will have leaved the gpu memory, so we should just release the underlaying java semaphore.

Or there exists some case i do not know. please correct me if wrong.

Thanks.

**Comment by revans2** [comment, marked as answer]
That code is checking to see if the current task owns the semaphore and releases it only if it does.

It is a little confusing because it uses a `MutableInt` as a reference count.  Originally it was intended that we would reference count how many times we asked to grab the semaphore and only release it when we hit the same number, but that proved to be impossible, so now the value of the reference count is always 1 or 0.  It goes to 1 when a task grabs the semaphore and goes to 0 when it releases the semaphore.  Al other times the code looks to see if the task already has the semaphore and lets it through, or the release becomes a noop.

**Comment by JustPlay** [comment]
> so now the value of the reference count is always 1 or 0

How ？

UPDATE：
https://github.com/NVIDIA/spark-rapids/blob/1fa70c3aa36833d751b07390959896f0a2924e74/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala#L111  
Oh, I know, i think this it is

**Comment by revans2** [comment]
Yes
